src/logics/Logic.cc

```diff
@@ -1600,3 +1605,98 @@ char* Logic::printTerm        (PTRef tr, bool l, bool s) const { return printTer
 void Logic::termSort(vec<PTRef>& v) const { sort(v, LessThan_PTRef()); }
 
 void  Logic::purify     (PTRef r, PTRef& p, lbool& sgn) const {p = r; sgn = l_True; while (isNot(p)) { sgn = sgn^1; p = getPterm(p)[0]; }}
+
+bool Logic::typeCheck(SymRef sym, vec<PTRef> const & args, std::string & why) const {
+
+    auto genSortMismatchString = [&](SymRef sym, vec<PTRef> const & args) {
+        std::string symStr = getSymName(sym);
+        Symbol const & symbol = sym_store[sym];
+        if (symbol.chainable() or symbol.pairwise()) {
+            for (int i = 0; i < args.size(); i++) {
+                symStr += " " + std::string(getSortName(symbol[0]));
+            }
+        } else if (symbol.left_assoc()) {
+            symStr += " " + std::string(getSortName(symbol[0]));
+            for (int i = 1; i < args.size(); i++) {
+                symStr += " " + std::string(getSortName(symbol[1]));
+            }
+        } else if (symbol.right_assoc()) {
+            for (int i = 0; i < args.size() - 1; i++) {
+                symStr += " " + std::string(getSortName(symbol[0]));
+            }
+            symStr += " " + std::string(getSortName(symbol[1]));
```

Comment:
We want the last argument of symbol, not the second one, no?
```suggestion
            symStr += " " + std::string(getSortName(symbol[symbol.size() - 1]));
```

Reply:
You're right, but the last element is always `symbol[1]`.  There's an internal invariant that the Left- and Right-Associative symbols must be binary in the sense that they have two arguments. This is not enforced anywhere though  For now I'll try to add an assert in `Symbol::set{Left,Right}Assoc()`.  

The `Symbol::setX()` methods should be methods in `SymbolStore`, and the chainable and pairwise arguments could be made unary with the same logic.   That's for another PR (that is yet to be written, not in my queue).

Reply:
Yeah, I missed that the symbol has to be binary in this case.
Makes sense.